pkg/zng/docs/spec.md

```diff
@@ -1,92 +1,366 @@
 
-Any line that begins with character `#` is a directive.
-All other lines are values.  If a value line begins with `#`, then this
-character must be escaped.
-(Such lines can only exist as legacy values and not regular values
-as regular values always begin with an integer descriptor.)
+Control code 1 is reserved for the ordering hint.  All other control codes
+are available to higher-layer protocols for carrying protocol-specific payloads
+embedded in the ZNG stream.
 
-Directives indicate how subsequent values in the ZSON stream are interpreted.
-A value is a regular value if the most recent preceding directive in the stream
-is a regular directive; otherwise, it is a legacy value.
+Control messages may be used informatively and shall be
+ignored by any data receivers.  The message can be any UTF-8 string.
+Control payloads are guaranteed to be preserved
+in order within the stream and presented to higher layer components through
+any ZNG streaming API.  In this way, senders and receivers of ZNG can embed
+protocol directives as ZNG control payloads rather than defining additional
+encapsulating protocols.  See the
+[zng-over-http](zng-over-http.md) protocol for an example.
+
+### BZNG Types
+
+Following a header byte of 0x80 is a "type definition" (typedef) that binds
+"the next available" integer type code to a type encoding.  Type codes
+begin at the value 23 and increase by one for each typedef. This bindings
+are scoped to the stream in which the typedef occurs.
+
+Type codes for the "scalar types" are predefined as follows:
+
+| Type     | Code |
+|----------|------|
+| bool     |   0  |
+| byte     |   1  |
+| int16    |   2  |
+| uint16   |   3  |
+| int32    |   4  |
+| uint32   |   5  |
+| int64    |   6  |
+| uint64   |   7  |
+| float64  |   8  |
+| string   |   9  |
+| bytes    |  10  |
+| bstring  |  11  |
+| enum     |  12  |
+| ip       |  13  |
+| port     |  14  |
+| net      |  15  |
+| time     |  16  |
+| duration |  17  |
+| any      |  18  |
+
+#### BZNG typedef
+
+A typedef defines a new composite type (i.e., record, array, or set)
+from the composite type code plus the composite type-specific encoding of the
+new type.
+
+The composite type codes are:
+
+| Type     | Code |
+|----------|------|
+| record   |   0  |
+| array    |   1  |
+| set      |   2  |
+
+Note that the composite type codes overlap with the scalar type codes because their
+uses never overlap.  Composite type codes can only appear at the beginning of
+a typedef.  Thus, the space of referenceable type codes in a given stream is
+the union of the scalar type codes and the type codes created with typedefs.
+
+A typedef is encoded as a single byte indicating the composite type code following by
+the type encoding.  This creates a binding between the implied type code
+(i.e., 23 plus the count of all previous typedefs in the stream) and the new
+type definition.
+
+A "uvarint" is an unsigned integer encoded according to the protobuf specification.
```

Comment:
This reference to uvarint seemed odd to me here on first read given it hadn't been used in any of the tables above. I see it comes up below of course, but maybe there's a way to introduce it better.

Also, this is me not knowing protobufs, but I assume we truly mean "uvarint" and not "varint"? I was going to suggest that we include a link here directly into the protocol spec to where they describe uvarint. My Google search actually changed over to just "varint" as if I'd made a typo, showing me 31k "varint" results. When I forced it to truly look for "uvarint", it showed me only 377 results, and the top hits were all in code, not in spec. So it's definitely out there, and maybe it's what we intended, but seems a little esoteric to me as an outsider.

Reply:
Good point.

Reply:
fixed